src/js/osweb/items/loop.js

```diff
@@ -182,85 +213,68 @@ export default class Loop extends Item {
-        this._cycles.push(i)
-      }
-    }
-
-    // Add the leftover repeats.
-    const partialRepeats = this.vars.get('repeat') - wholeRepeats
-    if (partialRepeats > 0) {
-      const allCycles = Array.apply(null, {
-        length: this.vars.cycles
-      }).map(Number.call, Number)
-      const remainder = Math.floor(this.vars.cycles * partialRepeats)
-      for (let i = 0; i < remainder; i++) {
-        // Calculate random position.
-        const position = Math.floor(Math.random() * allCycles.length)
-        // Add position to cycles.
-        this._cycles.push(position)
-        // Remove position from array.
-        allCycles.splice(position, 1)
-      }
-    }
-
-    // Randomize the list if necessary.
-    if (this.vars.get('order') === 'random') {
-      this._cycles = shuffle(this._cycles)
-    } else {
-      const skipVal = this.vars.get('skip')
-      // In sequential order, the offset and the skip are relevant.
-      if (this._cycles.length < skipVal) {
-        this.experiment._runner._debugger.addError('The value of skip is too high in loop item. ' +
-          'You cannot skip more cycles than there are in: ' + this.name)
-      } else {
-        if (this.vars.get('offset') === 'yes') {
-          // Get the skip elements.
-          const skip = this._cycles.slice(0, skipVal)
-
-          // Remove the skip elements from the original location.
-          this._cycles = this._cycles.slice(skipVal)
-
-          // Add the skip element to the end.
-          this._cycles = this._cycles.concat(skip)
-        } else {
-          this._cycles = this._cycles.slice(skipVal)
-        }
-      }
-    }
-
-    this._keyboard = new Keyboard(this.experiment)
 
     // Make sure the item to run exists.
     if (this.experiment.items._items[this.vars.item] === 'undefined') {
-      this.experiment._runner._debugger.addError('Could not find an item which is called by loop item: ' +
+      this._runner._debugger.addError('Could not find an item which is called by loop item: ' +
       this.name + ' (' + this.vars.item + ')')
     }
-
     super.prepare()
-
     this.set_item_onset()
   }
 
   /** Implements the run phase of an item. */
   run () {
     super.run()
 
+    if (!this._initialized) {
+      // Randomize the list if necessary.
+      if (this.vars.get('order') === 'random') {
+        this._cycles = shuffle(this._cycles)
+      }
+
+      // Perform the operations
+      this.matrix = this._operations.reduce((mtrx, [func, args]) =>
+        func(mtrx, ...this._eval_args(args)), this.matrix)
+      // Set the number of cycles to the length of the generated matrix
+      this.vars.cycles = this.matrix.length
+
+      // Walk through all complete repeats
+      const wholeRepeats = Math.floor(this.vars.get('repeat'))
+      for (let j = 0; j < wholeRepeats; j++) {
+        for (let i = 0; i < this.vars.cycles; i++) {
+          this._cycles.push(i)
+        }
+      }
+
+      // Add the leftover repeats.

```

Comment:
This whole section about partial repeats that follows doesn't seem to do anything. A variable `allCycles` is created but nothing is done with it anymore afterwards, so it seems useless. I couldn't distill from OpenSesame's  python code either what this section is supposed to do, or what the partial repeats mean.
@smathot can you maybe elaborate, or just confirm that this section can safely be removed?

Reply:
This section should take into account non-integer repeat values. So say that the table is like this

Reply:
Oh right! I didn't even know that was possible :). That should be easy enough to implement and can probably be done more efficiently than the current code.

Reply:
If it works like this, let's leave it for now though. (Even if it's not super elegant.)

Reply:
It actually doesn't work, I think. The code creates and processes a variable `allCycles` but doesn't do anything with it afterwards. Shall I still see if I can get this to work?

Reply:
:+1: And make sure to check the results against the behavior of OpenSesame. I think this implementation probably does work, actually, but that just the `allCycles` constant doesn't do anything. The action is in the `this._cycles.push()`.